<a href="https://colab.research.google.com/github/eunseochu/YB_4th/blob/main/%ED%8C%8C%EC%9D%B4%EC%8D%AC_%EB%A8%B8%EC%8B%A0%EB%9F%AC%EB%8B%9D_%EC%99%84%EB%B2%BD_%EA%B0%80%EC%9D%B4%EB%93%9C_%ED%8C%90%EB%8B%A4%EC%8A%A4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **4. 판다스**

### **1) 판다스란?**
파이썬에서 데이터 처리를 위해 존재하는 가장 인기 있는 라이브러리
- 대부분의 데이터 세트가 **2차원 데이터** (행x열)
  * 2차원 데이터는 인간이 가장 이해하기 쉬운 데이터 구조이자 효과적으로 데이터를 담을 수 있는 구조

#### **1-1) 판다스 vs 넘파이**
판다스의 많은 부분은 넘파이 기반으로 작성되었으나, 넘파이보다 훨씬 유연하고 편리하게 데이터 핸들링 가능
- 넘파이의 데이터 핸들링은 저수준 API가 대부분이기 때문에 편하지 않음
- RDBMS의 SQL이나 엑셀 시트만큼은 아니더라도 이에 버금가는 고수준 API제공
- 파이썬의 리스트, 컬렉션, 넘파이 등 내부 데이터뿐만 아니라 CSV 등의 파일을 쉽게 DataFrame으로 변경해 <br/> 데이터의 가공/분석을 편리하게 수행 가능

#### **1-2) 판다스의 핵심 개체: DataFrame**
판다스가 다루는 대부분의 영역은 DataFrame과 관련된 부분
- **DataFrame**: 여러 개의 행과 열로 이뤄진 2차원 데이터 <br/>
- DataFrame을 이해하기 위해서는 다른 중요 객체인 Index와 Series를 이해하는 것도 중요

**1-2-1) Index** <br/>
RDBMS의 PK처럼 개별 데이터를 고유하게 식별하는 Key 값
  - Series와 DataFrame 모두 Index를 Key값으로 가지고 있음
  - Series는 **column이 하나**뿐이고 DataFrame은 **column이 여러 개**인 데이터 구조체 <br/> 

**1-2-2) DataFrame과 Series** <br/>
DataFrame은 여러 개의 Sereis로 이루어졌음
- DataFrame을 개괄적으로 이해한 후 Index 및 Series 이해

---

### **2) 판다스 시작 - 파일을 DataFrame으로 로딩, 기본 API**
csv, tab과 같은 다양한 유형의 분리 문자로 column을 분리한 파일을 손쉽게 DataFrame으로 로딩

#### **2-1) DataFrame API**
다양한 포맷으로 된 파일을 DataFrame으로 로딩할 수 있는 편리한 API 제공 <br/>

**2-1-1) read_csv** <br/>
CSV(',') 파일 포맷 변환을 위한 API
- 어떤 필드 구분 문자(Delimeter) 기반의 파일 포맷도 DataFrame으로 변환 가능
  * 인자인 sep에 해당 구분 문자 사용, 생략 시 자동으로 콤마 할당
  * ex. read_csv('파일명', sep='\t')
- **filepath**: 가장 중요한 인자로, 로드하려는 데이터 파일의 경로를 포함한 파일명을 입력
- 나머지 인자는 지정하지 않으면 디폴트 값으로 할당

**2-1-2) read_table()와 read_fwf()** <br/>
- read_table(): 디폴트 필드 구분 문자가 탭('\t')으로, read_csv()와 기능상 큰 차이 존재 X
- read_fwf(): 고정 길이 기반(Fixed Width)의 column 포맷을 DataFrame으로 로딩 <br/>

#### **2-2) DataFrame 로딩**
kaggle의 타이타닉 탑승자 데이터 파일 사용 (https://www.kaggle.com/c/titanic/data)

**2-2-1) 타이타닉 데이터 로딩** <br/>
pd.read_csv()는 호출 시 파일명 인자로 들어온 파일을 로딩해 DataFrame 객체로 반환 
- 별다른 파라미터 지정이 없으면 파일의 **맨 처음 row를 column명으로 인지**하고 column으로 변환
- 맨 왼쪽에 파일에 기재되지 않은 데이터값들이 row 순으로 순차적으로 표시
  * 판다스의 Index 객체 값으로, 고유의 레코드를 식별하는 역할
  * 모든 DataFrame 내의 데이터는 **생성되는 순간 고유의 Index 값**을 가지게 됨

In [12]:
import pandas as pd

titanic_df = pd.read_csv('/content/train.csv')
print('titanic 변수 type:', type(titanic_df))
titanic_df

titanic 변수 type: <class 'pandas.core.frame.DataFrame'>


PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0              1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1              2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2              3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S
3              4         1       1  Futrelle, M...  female  35.0      1      0          113803  53.1000  C123        S
4              5         0       3  Allen, Mr. ...    male  35.0      0      0          373450   8.0500   NaN        S
..           ...       ...     ...             ...     ...   ...    ...    ...             ...      ...   ...      ...
886          887         0       2  Montvila, R...    male  27.0      0      0          211536  13.0000   NaN        S
887          888         1       1  Graham, Mis...  female  19.0      0      0          112053  30.0000   B42        S
888          889         0       3  Johnston, M...  female   NaN      1      2      W./C. 6607  23.4500   NaN        S
889          890         1       1  Behr, Mr. K...    male  26.0      0      0          111369  30.0000  C148        C
890          891         0       3  Dooley, Mr....    male  32.0      0      0          370376   7.7500   NaN        Q

[891 rows x 12 columns]

**2-2-2) head(): 일부 데이터만 표출** <br/>
모든 데이터를 표출하면 보기 불편할 뿐만 아니라 메모리도 많이 사용하므로 일부 데이터만 표출
- head(3)은 맨 앞 3개의 row를 반환
- Default 값은 5개

In [13]:
titanic_df.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S

**2-2-3) shape 변수: DataFrame의 행과 열 크기** <br/>
shape는 DataFrame의 행과 열을 튜플 형태로 반환

In [14]:
print('DataFrame 크기: ', titanic_df.shape)

DataFrame 크기:  (891, 12)


- 891개의 row와 12개의 column으로 구성

#### **2-3) 메타 데이터 조회**
DataFrame은 데이터뿐만 아니라 column의 타입, Null 데이터 개수, 데이터 분포도 등의 메타 데이터 조회 가능

**2-3-1) info()** <br/>
총 데이터 건수와 데이터 타입, Null 건수 조회 가능
- **RangeIndex**: DataFrame index의 범위, 즉 전체 row 수
- **Dtype**: column별 데이터 타입
- **Non-Null Count**: 몇 개의 데이터가 non-null인지 나타냄
- 마지막 줄에 전체 column의 타입을 요약

In [15]:
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


- 전체 데이터는 891개의 row이며 column 수는 12개
- Name은 object 타입으로, 문자열 타입으로 생각
- Age column 891개의 데이터 중 714개가 Null이 아니며 177개는 Null
- 2개 column이 float64, 5개 column이 int64, 5개 column이 object 타입

**2-3-2) describe()** <br/>
column별 **숫자형 데이터값의 n-percentile 분포도, 평균값, 최댓값, 최솟값**을 나타냄
- **오직 숫자형(int, float 등) column**의 분포도만 조사하며, 자동으로 object 타입의 column을 출력에서 제외 
  * Count: Not null인 데이터 건수
  * Mean: 전체 데이터의 평균값
  * Std: 표준편차
  * Min: 최솟값
  * Max: 최댓값
  * 25/50/75%: 25/50/75 percentile 값
- 해당 숫자 column이 **숫자형 카테고리 column인지 판단**할 수 있게 도와줌
  * 카테고리 column은 특정 범주에 속하는 값을 코드화한 column인데, 숫자로 표시할 수 있음



In [16]:
titanic_df.describe()

PassengerId    Survived      Pclass         Age       SibSp       Parch        Fare
count   891.000000  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean    446.000000    0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std     257.353842    0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min       1.000000    0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%     223.500000    0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%     446.000000    0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%     668.500000    1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max     891.000000    1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

- PassengerID: 승객 ID를 식별하는 column이므로 분석을 위한 의미 있는 속성 X
- Survived: min = 0, 25~75% = 0, max = 1이므로 0과 1로 이루어진 숫자형 카테고리 column
- Pclass column: min = 1, 25~75% = 2와 3, max = 3이므로 1, 2, 3으로 이루어진 숫자형 카테고리 column

**2-3-3) value_counts()** <br/>
Series 객체에 value_counts() 메서드를 호출 시 **해당 column값의 유형과 건수 확인** 가능
- 지정된 column의 데이터값 전수를 반환
- 데이터의 분포도를 확인하는 매우 유용한 함수이며, 자주 사용됨
  * Series 객체에서만 정의되어 있으며, DataFrame은 가지고 있지 않음
  * DataFrame의 [ ] 연산자 내부에 column명을 입력하면 **Series 형태로 특정 column 데이터 세트가 반환**


In [17]:
titanic_pclass = titanic_df['Pclass'] # [ ] 연산자 내부에 column명 입력 시 해당 column에 해당하는 Series 객체 반환
print(type(titanic_pclass))

<class 'pandas.core.series.Series'>


In [18]:
value_counts = titanic_df['Pclass'].value_counts()
print(value_counts) # 3이 491개, 1이 216개, 2가 184개

3    491
1    216
2    184
Name: Pclass, dtype: int64


#### **2-4) Series**
Index와 단 하나의 column으로 구성된 데이터 세트

**2-4-1) Series의 Index** <br/>
왼쪽의 0부터 시작하는 순차값이 Index며, 오른쪽은 Series의 해당 column의 데이터값
- 이는, DataFrame의 index와 동일한 index
- 모든 Series와 DataFrame은 index를 반드시 가진다

In [19]:
titanic_pclass.head() # index, 데이터 값

0    3
1    1
2    3
3    1
4    3
Name: Pclass, dtype: int64

**2-4-2) value_counts()** <br/>
 Series 객체에서만 정의되며, 반환되는 타입 역시 Series 객체
- DataFrame은 value_counts() 메서드를 가지고 있지 않음
- value_counts()는 column 값별 데이터 건수를 반환하므로 고유 column 값을 식별자로 사용 가능

In [20]:
value_counts = titanic_df['Pclass'].value_counts()
print(type(value_counts)) # value_counts: Series 객체 반환
print(value_counts)

<class 'pandas.core.series.Series'>
3    491
1    216
2    184
Name: Pclass, dtype: int64


**2-4-3) Index의 특징** <br/>
순차 값과 같은 의미 없는 식별자만 할당하는 것이 아니라, 의미 있는 데이터값 할당도 가능 
- DataFrame, Series가 만들어진 후에도 변경 가능
- 숫자형뿐만 아니라 문자열도 가능
- 단, **고유성**이 보장되어야 함

---

### **3) DataFrame과 리스트, 딕셔너리, 넘파이 ndarray 상호 변환**
DataFrame은 파이썬의 리스트, 딕셔너리, 그리고 넘파이 ndarray 등 다양한 데이터로부터 생성될 수 있으며, <br/> 또한 반대로 파이썬의 리스트, 딕셔너리, 그리고 넘파이 ndarray 등으로 변환될 수 있음
- 기본적으로 **넘파이 ndarray를 입력 인자로 사용**하는 경우가 대부분
- DataFrame과 넘파이 ndarray 상호 간의 변환이 매우 빈번하게 발생

#### **3-1) 넘파이 ndarray, 리스트, 딕셔너리를 DataFrame으로 변환하기**
DataFrame은 리스트와 넘파이 ndarray와 다르게 column명을 가지고 있기에 편하게 데이터 핸들링 가능
- 일반적으로 DataFrame으로 변환 시 column명 지정 
  * 따로 지정하지 않으면 자동으로 column명을 할당
- DataFrame 객체의 생성인자 data는 리스트나 딕셔너리, 넘파이 ndarray를 입력 받음
- DataFrame 객체의 **생성인자 columns는 column명 리스트를 입력 받아** 쉽게 DataFrame 생성

**3-1-1) 1차원 형태의 리스트와 넘파이 ndaray를 DataFrame으로 변환** <br/>
DataFrame은 2차원 데이터이기 때문에, 2차원 이하의 데이터들만 DataFrame으로 변환 가능
- 1차원 데이터이므로 **column명이 하나만 필요**

In [21]:
import numpy as np

col_name1 = ['col1']
list1 = [1, 2, 3]
array1 = np.array(list1)
print('array1 shape:', array1.shape)

array1 shape: (3,)


In [22]:
# 리스트를 이용해 DataFrame 생성
df_list1 = pd.DataFrame(list1, columns = col_name1)
print('1차원 리스트로 만든 DataFrame:\n', df_list1)

# 넘파이 ndarray를 이용해 DataFrame 생성
df_array1 = pd.DataFrame(array1, columns = col_name1)
print('1차원 ndarray로 만든 DataFrame:\n', df_array1)

1차원 리스트로 만든 DataFrame:
    col1
0     1
1     2
2     3
1차원 ndarray로 만든 DataFrame:
    col1
0     1
1     2
2     3


**3-1-2) 2차원 형태의 리스트와 넘파이 ndaray를 DataFrame으로 변환** <br/>
2행 3열 형태의 리스트와 ndarray를 기반으로 DataFrame을 생성하므로 column명은 3개가 필요

In [23]:
# 3개의 column명이 필요함
col_name2 = ['col1', 'col2', 'col3']

# 2행x3열 형태의 리스트와 ndarray 생성한 뒤 이를 DataFrame으로 변환.
list2 = [[1, 2, 3],
         [11, 12, 13]]
array2 = np.array(list2)
print('array2 shape:', array2.shape)

array2 shape: (2, 3)


In [24]:
df_list2 = pd.DataFrame(list2, columns = col_name2)
print('2차원 리스트로 만든 DataFrame:\n', df_list2)
df_array2 = pd.DataFrame(array2, columns = col_name2)
print('2차원 ndarray로 만든 Dataframe:\n', df_array2)

2차원 리스트로 만든 DataFrame:
    col1  col2  col3
0     1     2     3
1    11    12    13
2차원 ndarray로 만든 Dataframe:
    col1  col2  col3
0     1     2     3
1    11    12    13


**3-1-3) 딕셔너리를 DataFrame으로 변환** <br/>
딕셔너리의 키(key)는 column명으로, 값(Value)는 key에 해당하는 column 데이터로 변환
- key의 경우는 문자열, 값의 경우 리스트(또는 ndarray) 형태로 딕셔너리 구성

In [25]:
# Key는 문자열 column명으로 매핑, Value는 리스트 형(또는 ndarray) column 데이터로 매핑
dict = {'col1': [1, 11], 'col2': [2, 22], 'col3': [3, 33]}
df_dict = pd.DataFrame(dict)
print('딕셔너리로 만든 DataFrame:\n', df_dict)

딕셔너리로 만든 DataFrame:
    col1  col2  col3
0     1     2     3
1    11    22    33


#### **3-2) DataFrame을 넘파이 ndarray, 리스트, 딕셔너리로 변환하기**

**3-2-1) DataFrame을 넘파이 ndarray로 변환** <br/>
데이터 핸들링은 DataFrame을 이용하더라도, 패키지 입력 인자 등에 적용하기 위해 다시 넘파이 ndarray로 변환
- 많은 머신러닝 패키지가 기본 데이터 형으로 넘파이 ndarray를 사용
- **DataFrame 객체의 values**를 이용해 쉽게 가능하며, 매우 많이 사용되는 방법

In [26]:
# DataFrame을 ndarray로 변환
array3 = df_dict.values
print('df_dict.values 타입: ', type(array3))
print('df_dict.values shape:', array3.shape)
print(array3)

df_dict.values 타입:  <class 'numpy.ndarray'>
df_dict.values shape: (2, 3)
[[ 1  2  3]
 [11 22 33]]


**3-2-2) DataFrame을 리스트와 딕셔너리로 변환**
- **리스트**
  * values로 얻은 **ndarray에 tolist()** 호출
- **딕셔너리**
  * **DataFrame 객체의 to_dict()** 메서드 호출
  * 인자로 'list' 입력 시 딕셔너리의 값이 리스트형으로 반환

In [27]:
# DataFrame을 리스트로 변환
list3 = df_dict.values.tolist()
print('df_dict.values.tolist() 타입:', type(list3))
print(list3)

df_dict.values.tolist() 타입: <class 'list'>
[[1, 2, 3], [11, 22, 33]]


In [28]:
# DataFrame을 딕셔너리로 변환
dict3 = df_dict.to_dict('list')
print('df_dict.values.tolist() 타입:', type(dict3))
print(dict3)

df_dict.values.tolist() 타입: <class 'dict'>
{'col1': [1, 11], 'col2': [2, 22], 'col3': [3, 33]}


---

### **4) DataFrame의 column 데이터 세트 생성과 수정**
DataFrame의 column 데이터 세트 생성과 수정 역시 **[ ] 연산자를 이용**해 쉽게 가능

#### **4-1) column 데이터 생성**

**4-1-1) 새로운 column 데이터 추가** <br/>
DataFrame [ ] 내에 새로운 column 명을 입력하고 값을 할당
- Series에 상숫값을 할당하면 Series의 모든 데이터 세트에 일괄 적용됨 (넘파이와 동일)

In [29]:
titanic_df['Age_0'] = 0 
titanic_df.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked  Age_0
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S      0
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C      0
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S      0

- 새로운 column명 'Age_0'으로 모든 데이터값이 0으로 할당된 Series가 기존 DataFrame에 추가

**6-1-2) 기존 column Sereis 데이터를 이용한 새로운 column Series 생성**

In [30]:
titanic_df['Age_by_10'] = titanic_df['Age']*10
titanic_df['Family_No'] = titanic_df['SibSp'] + titanic_df['Parch'] + 1
titanic_df.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked  Age_0  Age_by_10  Family_No
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S      0      220.0          2
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C      0      380.0          2
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S      0      260.0          1

#### **4-2) 기존 column 값 일괄 업데이트**
업데이트를 원하는 column Series를 DataFrame[ ] 내에 column 명으로 입력한 뒤 값을 할당

In [31]:
titanic_df['Age_by_10'] = titanic_df['Age_by_10'] + 100
titanic_df.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked  Age_0  Age_by_10  Family_No
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S      0      320.0          2
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C      0      480.0          2
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S      0      360.0          1

- Age_by_10 column의 모든 값이 기존 값 + 100 값으로 업데이트 됨

### **5) DataFrame 데이터 삭제**
DataFrame에서 데이터 삭제는 drop() 메서드 사용
- 사용하는 데 혼동을 줄 만한 부분이 있어 주의 필요

#### **5-1) drop() 메서드**
labels, axis, inplace 파라미터가 가장 중요
```
DataFrame.drop(labels = None, axis = 0, index = None, columns = None, level = None, inplace = False, erros = 'raise')
```

**5-1-1) axis** <br/>
axis값에 따라 특정 column 또는 특정 row 드롭 
- **axis 0 = row 방향 축, axis 1 = column 방향 축** 
- DataFrame은 2차원 데이터이므로 axis 0, 1로 구성

**5-1-2) labels** <br/>
- labels에 **원하는 column명을 입력**하고 axis = 1을 입력하면 지정된 column을 드롭
  * 대부분의 경우는 column을 drop하는 경우
  * 기존 column 값을 가공해 새로운 column을 만들고 삭제하는 경우가 많음
- axis = 0으로 지정하면 자동으로 **labels에 오는 값을 index로 간주**
  * DataFrame의 특정 row를 가리키는 것은 index
  * row 레벨로 삭제하는 경우는 이상치 데이터를 삭제할 때 주로 사용

In [32]:
titanic_drop_df = titanic_df.drop('Age_0', axis = 1)
titanic_drop_df.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked  Age_by_10  Family_No
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S      320.0          2
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C      480.0          2
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S      360.0          1

- drop을 수행한 결과가 titanic_drop_df 변수로 반환됨
- axis = 1로 지정했기에 'Age_0' column이 삭제됨

**5-1-3) inplace** <br/>
**inplace = False**면 **자기 자신의 DataFrame 데이터는 삭제하지 않으며**, 삭제된 결과 **DataFrame을 반환**
- Default 값이 False이므로 inplace 파라미터 미기재 시 자동으로 False가 됨
- True로 설정하면 자신의 DataFrame 데이터를 삭제

In [33]:
titanic_df.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked  Age_0  Age_by_10  Family_No
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S      0      320.0          2
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C      0      480.0          2
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S      0      360.0          1

#### **5-2) drop() 응용**

**5-2-1) 여러 개의 column 삭제** <br/>
리스트 형태로 삭제하고자 하는 column 명을 입력해 labels 파라미터로 입력

In [34]:
drop_result = titanic_df.drop(['Age_0', 'Age_by_10', 'Family_No'], axis = 1, inplace = True)
print('inplace = True로 drop 후 반환된 값:', drop_result)
titanic_df.head(3)

inplace = True로 drop 후 반환된 값: None


PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S

- 'Age_0', 'Age_by_10', 'Family_No' 세 개의 column이 모두 삭제
- inplace를 True로 설정 시 반환 값이 None이 됨
  * 반환 값을 다시 자신의 DataFrame 객체로 할당하면 안됨
  * DataFrame 객체 변수를 아예 None으로 만들어 버림
  * ex. titanic_df = titanic_df.drop(['Age_0', 'Age_by_10', 'Family_No'], axis = 1, inplace = True)

**5-2-2) row 삭제** <br/>
axis = 0으로 설정하고 Index 0, 1, 2 row 삭제

In [35]:
pd.set_option('display.width', 1000)
pd.set_option('display.max_colwidth', 15)

print('#### before axis 0 drop ####')
print(titanic_df.head(3))
titanic_df.drop([0, 1, 2], axis = 0, inplace = True)

print('#### after axis 0 drop ####')
print(titanic_df.head(3))

#### before axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S
#### after axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
3            4         1       1  Futrelle, M...  female  35.0      1      0  113803  53.1000  C123        S
4            5         0       3  Allen, Mr. ...    male  35.0      0      0  373450   8.0500   NaN        S
5            6         0       3  Moran, Mr. ...    male   NaN      0      0  330877   8.4583   NaN        Q


- 인덱스 0, 1, 2에 위치한 row가 삭제됨

#### **5-3) drop() 메서드로 DataFrame 변경하기**
- DataFrame의 row를 삭제할 때는 axis = 0, column을 삭제할 때는 axis = 1로 설정
- 원본 DataFrame은 유지하고 드롭된 DataFrame을 **새롭게 객체 변수로 받고 싶다면 inplace = False**로 설정 (Default)
- 원본 DataFrame에 **드롭된 결과를 적용할 경우에는 inplace = True**를 적용 <br/>
- 드롭된 DataFrame을 다시 원본 DataFrame 객체 변수로 할당하면, 원본 DataFrame에서 드롭된 결과를 적용할 경우와 같음 <br/>
  * 단, 기존 원본 DataFrame 객체 변수는 메모리에서 추후 제거됨
  * ex. titanic_df = titanic_df.drop('Age_0', axis = 1, inplace = False)

### **6) Index 객체**
DataFrame, Series의 레코드를 고유하게 식별하는 객체
- DataFrame, Series에서 Index 객체만 추출하려면 **DataFrame.index 또는 Series.index 속성** 사용

#### **6-1) Index 객체 추출** <br/>
titanic_df DataFrame에서 Index 객체 추출
- drop으로 원본 Titanic DataFrame의 일부 데이터가 삭제되었으므로 다시 로딩 후 수행

**6-1-1) Index 객체의 특징** <br/>
Index 객체는 식별성 데이터를 1차원 array로 가지고 있음
- Index 객체의 실제 값은 넘파이 1차원 ndarray로 볼 수 있음
- values 속성으로 ndarray 값 확인 가능

In [36]:
# 원본 파일 다시 로딩
titanic_df = pd.read_csv('/content/train.csv') 

# Index 객체 추출
indexes = titanic_df.index
print(indexes)

RangeIndex(start=0, stop=891, step=1)


In [37]:
# Index 객체를 실제 값 array로 변환
print('Index 객체 array값:\n', indexes.values)

Index 객체 array값:
 [  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17
  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35
  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53
  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71
  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89
  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143
 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161
 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179
 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197
 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215
 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232 233
 234 235 236 237 238 239 240 241 

In [38]:
print(type(indexes.values))
print(indexes.values.shape)

<class 'numpy.ndarray'>
(891,)


**6-1-2) Index 단일 값 반환 및 슬라이싱** <br/>
ndarray와 유사하게 단일 값 반환 및 슬라이싱 가능

In [39]:
print(indexes[:5].values)
print(indexes.values[:5])
print(indexes[6])

[0 1 2 3 4]
[0 1 2 3 4]
6


**6-1-3) Index 객체 변경** <br/>
한 번 만들어진 DataFrame 및 Series의 Index 객체는 함부로 변경할 수 없음
- 첫 번째 Index 객체의 값을 5로 변경하는 작업은 수행할 수 없어 오류

In [40]:
indexes[0] = 5

TypeError: ignored

#### **6-2) Series 객체의 index**
Series 객체에 연산 함수를 적용할 때 Index는 연산에서 제외됨
- Series 객체는 1차원 배열과 같은 자료구조지만, 리스트와 달리 **인덱싱 값을 지정**할 수 있음
- Series 객체는 Index 객체를 포함하지만, Index는 오직 식별용으로만 사용

In [41]:
series_fair = titanic_df['Fare']
series_fair

0       7.2500
1      71.2833
2       7.9250
3      53.1000
4       8.0500
        ...   
886    13.0000
887    30.0000
888    23.4500
889    30.0000
890     7.7500
Name: Fare, Length: 891, dtype: float64

In [42]:
print('Fair Series max 값:', series_fair.max())
print('Fair Series sum 값:', series_fair.sum())
print('sum() Fair Series:', sum(series_fair))
print('Fair Series + 3:\n', (series_fair + 3).head(3))

Fair Series max 값: 512.3292
Fair Series sum 값: 28693.9493
sum() Fair Series: 28693.949299999967
Fair Series + 3:
 0    10.2500
1    74.2833
2    10.9250
Name: Fare, dtype: float64


#### **6-3) reset_index() 메서드**
DataFrame 및 Series에 reset_index() 메서드를 수행하면 **새롭게 인덱스를 연속 숫자 형으로 할당**
- 기존 인덱스는 'index'라는 새로운 column 명으로 추가

In [43]:
titanic_reset_df = titanic_df.reset_index(inplace = False)
titanic_reset_df.head(3)

index  PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0      0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1      1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2      2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S

**6-3-1) reset_index() 활용** <br/>
인덱스가 연속된 int 숫자형 데이터가 아닐 경우에 다시 연속 int 숫자형 데이터로 만들 때 주로 사용
- value_counts() 수행 시 column의 고유 값이 식별자 인덱스 역할을 하는 경우 등

**6-3-2) Series에 reset_index() 적용** <br/>
Series에 reset_index() 적용 시 Series가 아닌 DataFrame이 반환됨
- 새롭게 연속 숫자형 인덱스가 만들어지고 기존 인덱스가 column으로 추가되어 **column이 2개**가 되기 때문
- **drop = TRUE**로 설정하면 기존 인덱스는 새로운 column으로 추가되지 않고 삭제
  - 새로운 column이 추가되지 않고 삭제되므로 그대로 Series로 유지

In [44]:
print('### before reset_index ###')
value_counts = titanic_df['Pclass'].value_counts()
print(value_counts)
print('value_counts 객체 변수 타입:', type(value_counts))

### before reset_index ###
3    491
1    216
2    184
Name: Pclass, dtype: int64
value_counts 객체 변수 타입: <class 'pandas.core.series.Series'>


In [45]:
new_value_counts = value_counts.reset_index(inplace = False)
print('### after reset_index ###')
print(new_value_counts)
print('new_value_counts 객체 변수 타입:', type(new_value_counts))

### after reset_index ###
   index  Pclass
0      3     491
1      1     216
2      2     184
new_value_counts 객체 변수 타입: <class 'pandas.core.frame.DataFrame'>


### **7) 데이터 셀렉션 및 필터링**
판다스의 데이터 셀렉션 및 필터링은 넘파이와 유사하면서도 다르기 때문에 혼동하기 쉬움
- **넘파이**: [ ] 연산자 내 단일 값 추출, 슬라이싱, 팬시 인덱싱, 불린 인덱싱을 통해 데이터 추출
  * 넘파이의 데이터 핸들링은 데이터 분석용으로 사용하기에는 편의성이 떨어짐
- **판다스**: ix[ ], iloc[ ], loc[ ] 연산자를 통해 동일 작업 수행
  * 판다스는 넘파이 기능을 계승하기도, 완전히 다른 기능을 제공하기도 함
  * 판다스의 DataFrame과 Series 끼리도 데이터 셀렉션 기능이 달라지기도 함

#### **7-1) DataFrame의 [ ] 연산자**
넘파이와 DataFrame 간 데이터 셀렉션에서 가장 유의해야할 부분은 '[ ]' 연산자
- 넘파이에서 [ ] 연산자는 행의 위치, 열의 위치, 슬라이싱 범위 등을 지정해 데이터를 가져올 수 있음
- DataFrame의 [ ] 안에 들어갈 수 있는 것은 **column 명 문자(또는 column 명의 리스트 객체), 또는 index로 변환 가능한 표현식**
  - DataFrame 뒤에 있는 [ ]는 **column만 지정할 수 있는 'column 지정 연산자'**

**7-1-1) DataFrame의 [ ] 내에 숫자 값을 입력할 경우** <br/>
**DataFrame 뒤의 [ ]에 column 명을 지정하지 않는 경우** 오류 발생
- titanic_df[0]과 같은 표현식은 0이 column 명이 아니기 때문에 오류 발생
- titanic_df[0, 0]이나 titanic_df[[0, 1, 2]]도 오류 발생

In [75]:
print('단일 column 데이터 추출:\n', titanic_df['Pclass'].head(3))
print('\n여러 column의 데이터 추출:\n', titanic_df[['Survived', 'Pclass']].head(3))

단일 column 데이터 추출:
 0    3
1    1
2    3
Name: Pclass, dtype: int64

여러 column의 데이터 추출:
    Survived  Pclass
0         0       3
1         1       1
2         1       3


In [47]:
print('[ ] 안에 숫자 index는 KeyError 오류 발생:\n', titanic_df[0]) # 0은 column명이 아니기에 오류 발생

KeyError: ignored

**7-1-2) DataFrame의 [ ] 내에 슬라이싱을 이용할 경우** <br/>
판다스의 **인덱스 형태로 변환 가능한 표현식**은 [ ]안에 입력 가능
- 슬라이싱을 이용하면 정확히 원하는 결과 반환

In [48]:
titanic_df[0:2]

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch     Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0  A/5 21171   7.2500   NaN        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0   PC 17599  71.2833   C85        C

**7-1-3) DataFrame의 [ ] 내에 불린 인덱싱 표현을 사용할 경우** <br/>
[ ] 내의 불린 인덱싱 기능을 통해 원하는 데이터를 편리하게 추출 가능
- 대부분의 데이터 셀렉션은 불린 인덱싱을 기반으로 함

In [49]:
titanic_df[titanic_df['Pclass'] == 3].head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket   Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171  7.250   NaN        S
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...  7.925   NaN        S
4            5         0       3  Allen, Mr. ...    male  35.0      0      0          373450  8.050   NaN        S

**7-1-4) DataFrame의 [ ] 연산자의 입력 인자 정리** <br/>
- DataFrame 바로 뒤의 [ ] 연산자는 넘파이나 Series의 [ ]와 다르다
- [  ] 내 입력 값은 column명(또는 column의 리스트)을 지정해 **column 지정 연산에 사용하거나 불린 인덱스 용도**로만 사용
- DataFrame[0:2]와 같은 슬라이싱 연산으로 데이터를 추출하는 방법은 사용하지 않는 것을 권장

#### **7-2) DataFrame ix[ ] 연산자**
**행 위치 지정으로 인덱스값, 열 위치 지정으로 column 명**을 입력해 원하는 위치의 데이터 추출 가능
- 행에 해당하는 위치 지정은 DataFrame의 인덱스값을 입력
- 열에 해당하는 위치 지정은 column 명뿐만 아니라 column의 위치 값 지정도 가능
- 넘파이 ndarray의 [ ] 연산자와 유사한 기능 제공 
  - column 명칭(label) 기반 인덱싱과 column 위치(position) 기반 인덱싱 모두 제공
  - 단일 지정, 슬라이싱, 불린 인덱싱, 팬시 인덱싱 모두 가능
- 코드가 혼란을 주거나 가독성이 떨어지게 되어 ix[ ]는 판다스에서 삭제 됨
  - column 명칭 기반 인덱싱 연산자인 loc[ ]와 column 위치 기반 인덱싱 연산자인 iloc[ ] 탄생
  - ix[ ] 사용 시 DeprecationWarning 출력

#### **7-3) 명칭 기반 인덱싱과 위치 기반 인덱싱의 구분** <br/>
- **명칭(Label) 기반 인덱싱**
  - **Column의 명칭을 기반으로** 위치를 지정하는 방식 
  - 'Column 명' 같이 명칭으로 열 위치를 지정하는 방식
- **위치(Position) 기반 인덱싱**
  - 0을 출발점으로 하는 가로축, 세로축 좌표 기반의 행과 열 위치를 기반으로 데이터 지정
  - **행, 열 값으로 정수 입력**

#### **7-4) DataFrame iloc[ ] 연산자**
**위치 기반 인덱싱**만 허용하기 때문에 행과 열 값으로 **integer 또는 integer형의 슬라이싱, 팬시 리스트 값**을 입력 <br/>
- 위치 인덱싱이 아닌 명칭을 입력하거나 문자열 인덱스를 행 위치에 입력할 경우 오류 발생
- 불린 인덱싱은 조건을 기술하므로 이에 제약받지 않음

In [50]:
data = {'Name': ['Chulmin', 'Eunkyung','Jinwoong','Soobeom'],
        'Year': [2011, 2016, 2015, 2015],
        'Gender': ['Male', 'Female', 'Male', 'Male']
       }
data_df = pd.DataFrame(data, index=['one','two','three','four'])
data_df

Name  Year  Gender
one     Chulmin  2011    Male
two    Eunkyung  2016  Female
three  Jinwoong  2015    Male
four    Soobeom  2015    Male

In [51]:
# data_df를 reset_index()로 새로운 숫자형 인덱스를 생성
data_df_reset = data_df.reset_index()
data_df_reset = data_df_reset.rename(columns = {'index':'old_index'})

# 인덱스값에 1을 더해서 1부터 시작하는 새로운 인덱스값 생성
data_df_reset.index = data_df_reset.index + 1
data_df_reset

old_index      Name  Year  Gender
1       one   Chulmin  2011    Male
2       two  Eunkyung  2016  Female
3     three  Jinwoong  2015    Male
4      four   Soobeom  2015    Male

In [52]:
data_df.iloc[0, 0]

'Chulmin'

In [53]:
# 다음 코드는 오류를 발생합니다.
data_df.iloc[0, 'Name'] 

ValueError: ignored

- iloc[ ]에 위치 인덱싱이 아닌 명칭 입력 시 오류

In [54]:
# 다음 코드는 오류를 발생합니다.
data_df.iloc['one', 0] 

ValueError: ignored

- iloc[ ]에 문자열 인덱스를 행 위치에 입력 시 오류

In [55]:
data_df_reset.iloc[0, 1] 

'Chulmin'

- 명확히 위치 기반 인덱싱이므로 첫 번째 행 두번째 열의 위치 값 반환

#### **7-5) DataFrame loc[ ] 연산자** <br/>
loc[ ]는 명칭 기반으로 데이터를 추출하기 때문에 **행 위치에는 DataFrame index 값을, 열 위치에는 column 명을 입력**

In [56]:
data_df

Name  Year  Gender
one     Chulmin  2011    Male
two    Eunkyung  2016  Female
three  Jinwoong  2015    Male
four    Soobeom  2015    Male

In [57]:
data_df.loc['one', 'Name']

'Chulmin'

- 인덱스값이 'one'인 행의 column 명이 'name'인 데이터 추출

**7-5-1) loc[ ]에 숫자형 index 입력** <br/>
index가 숫자형일 수 있기 때문에 문자열을 무조건 입력한다는 선입견을 가져서는 안됨

In [58]:
data_df_reset.loc[1, 'Name']

'Chulmin'

In [59]:
# 다음 코드는 오류를 발생합니다.
data_df_reset.loc[0, 'Name']

KeyError: ignored

- 인덱스 값이 0인 행이 없어 오류를 반환

**7-5-2) loc[ ]에 슬라이싱 적용** <br/>
일반적으로 슬라이싱을 '시작 값:종료 값'으로 지정 시 시작 값 ~ 종료 값 -1 까지의 범위를 의미하지만, <br/> loc[ ]의 경우 **종료값 -1이 아닌 종료값까지 포함**하는 것을 의미
- 명칭은 숫자 형이 아닐 수 있기 때문에 -1을 할 수 없음

In [60]:
data_df

Name  Year  Gender
one     Chulmin  2011    Male
two    Eunkyung  2016  Female
three  Jinwoong  2015    Male
four    Soobeom  2015    Male

In [62]:
print('위치 기반 iloc slicing\n', data_df.iloc[0:1, 0], '\n')
print('명칭 기반 loc slicing\n', data_df.loc['one':'two', 'Name'], '\n')

위치 기반 iloc slicing
 one    Chulmin
Name: Name, dtype: object 

명칭 기반 loc slicing
 one     Chulmin
two    Eunkyung
Name: Name, dtype: object 



- 위치 기반 iloc[0:1, 0]은 0번째 행 위치에 해당하는 1개의 행 반환
- 명칭 기반 loc[ ]을 적용한 경우에는 2개의 행 반환

**7-5-3) loc[ ]에 정수형 인덱스 사용** <br/>
DataFrame의 인덱스가 정수형이면 loc[ ]를 이용할 때 조심해야 함
- 1:2는 DataFrame의 인덱스값에 따른 명칭 기반 인덱싱이므로 인덱스가 1, 2인 2개의 데이터 반환

In [63]:
print(data_df_reset.loc[1:2, 'Name']) 

1     Chulmin
2    Eunkyung
Name: Name, dtype: object


#### **7-6) iloc, loc의 문제점과 주의할 점 정리**
- 명칭 기반 인덱싱과 위치 기반 인덱싱의 차이
  - DataFrame의 **인덱스나 column명으로 데이터에 접근**하는 것은 명칭 기반 인덱싱
  - 0부터 시작하는 **행, 열의 위치 좌표에만 의존**하는 것은 위치 기반 인덱싱
- **위치 기반 인덱싱 iloc[ ]**
  - **행과 열 위치 값으로 정수형 값**을 지정해 원하는 데이터를 반환
- **명칭 기반 인덱싱 loc[ ]**
  - **행 위치에 DataFrame 인덱스, 열 위치에는 Column명**을 지정해 원하는 데이터 반환
  - 슬라이싱을 '시작점:종료점'으로 지정할 때 시작점에서 **종료점을 포함한** 위치에 있는 데이터를 반환

#### **7-7) 불린 인덱싱**
매우 편리한 데이터 필터링 방식으로, 명확히 인덱싱을 지정하는 방식보다 많이 사용
- [ ], loc[ ]에서 공통적으로 지원
- iloc[ ]는 정수형 값이 아닌 불린 값에 대해서는 지원하지 않기 때문에 불린 인덱싱이 지원되지 않음

**7-7-1) 불린 인덱싱 예시** <br/>
[ ] 연산자 내 불린 조건을 입력하면 불린 인덱싱으로 자동으로 결과를 찾아줌

In [64]:
titanic_df = pd.read_csv('/content/train.csv') 
titanic_boolean = titanic_df[titanic_df['Age'] > 60]
titanic_boolean

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch       Ticket      Fare        Cabin Embarked
33            34         0       2  Wheadon, Mr...    male  66.0      0      0   C.A. 24579   10.5000          NaN        S
54            55         0       1  Ostby, Mr. ...    male  65.0      0      1       113509   61.9792          B30        C
96            97         0       1  Goldschmidt...    male  71.0      0      0     PC 17754   34.6542           A5        C
116          117         0       3  Connors, Mr...    male  70.5      0      0       370369    7.7500          NaN        Q
170          171         0       1  Van der hoe...    male  61.0      0      0       111240   33.5000          B19        S
252          253         0       1  Stead, Mr. ...    male  62.0      0      0       113514   26.5500          C87        S
275          276         1       1  Andrews, Mi...  female  63.0      1      0        13502   77.9583           D7        S
280          281         0       3  Duane, Mr. ...    male  65.0      0      0       336439    7.7500          NaN        Q
326          327         0       3  Nysveen, Mr...    male  61.0      0      0       345364    6.2375          NaN        S
438          439         0       1  Fortune, Mr...    male  64.0      1      4        19950  263.0000  C23 C25 C27        S
456          457         0       1  Millet, Mr....    male  65.0      0      0        13509   26.5500          E38        S
483          484         1       3  Turkula, Mr...  female  63.0      0      0         4134    9.5875          NaN        S
493          494         0       1  Artagaveyti...    male  71.0      0      0     PC 17609   49.5042          NaN        C
545          546         0       1  Nicholson, ...    male  64.0      0      0          693   26.0000          NaN        S
555          556         0       1  Wright, Mr....    male  62.0      0      0       113807   26.5500          NaN        S
570          571         1       2  Harris, Mr....    male  62.0      0      0  S.W./PP 752   10.5000          NaN        S
625          626         0       1  Sutton, Mr....    male  61.0      0      0        36963   32.3208          D50        S
630          631         1       1  Barkworth, ...    male  80.0      0      0        27042   30.0000          A23        S
672          673         0       2  Mitchell, M...    male  70.0      0      0   C.A. 24580   10.5000          NaN        S
745          746         0       1  Crosby, Cap...    male  70.0      1      1    WE/P 5735   71.0000          B22        S
829          830         1       1  Stone, Mrs....  female  62.0      0      0       113572   80.0000          B28      NaN
851          852         0       3  Svensson, M...    male  74.0      0      0       347060    7.7750          NaN        S

In [65]:
print(type(titanic_boolean))

<class 'pandas.core.frame.DataFrame'>


- titanic_df 'Age' column 값이 60보다 큰 데이터 모두 반환
- 반환된 titnaic_boolean 객체의 타입은 DataFrmae

**7-7-2) 불린 인덱싱으로 원하는 column 명 별도 추출** <br/>
[ ] 내에 불린 인덱싱을 적용하면 반환되는 객체가 DataFrame이므로 원하는 column 명만 별도 추출 가능
- [  ] 연산자에 column 명 입력하며, **column이 두 개 이상이므로 [[ ]]를 사용**

In [66]:
titanic_df[titanic_df['Age'] > 60][['Name', 'Age']].head(3)

Name   Age
33  Wheadon, Mr...  66.0
54  Ostby, Mr. ...  65.0
96  Goldschmidt...  71.0

In [67]:
titanic_df.loc[titanic_df['Age'] > 60, ['Name', 'Age']].head(3)

Name   Age
33  Wheadon, Mr...  66.0
54  Ostby, Mr. ...  65.0
96  Goldschmidt...  71.0

- loc[ ]을 이용해도 동일하게 적용 가능하지만, [ 'Name', 'Age' ] 는 column 위치에 놓아야 함

**7-7-3) 복합조건을 결합한 불린 인덱싱** <br/>
여러 개의 복합 조건도 결합해 적용 가능 <br/>
- 개별 조건은 ()로 묶고, 복합 조건 연산자를 사용
- and 조건일 때는 &, or 조건일 때는 |, Not 조건일 때는 ~

In [68]:
titanic_df[(titanic_df['Age'] > 60) & (titanic_df['Pclass'] == 1) & (titanic_df['Sex'] == 'female')]

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
275          276         1       1  Andrews, Mi...  female  63.0      1      0   13502  77.9583    D7        S
829          830         1       1  Stone, Mrs....  female  62.0      0      0  113572  80.0000   B28      NaN

- 나이가 60세 이상이고, 선실 등급이 1등급이며, 성별이 여성인 승객 추출

**7-7-4) 조건 변수를 이용한 불린 인덱싱** <br/>
개별 조건을 변수에 할당하고 이들 변수를 결합해 불린 인덱싱 수행 가능

In [69]:
cond1 = titanic_df['Age'] > 60
cond2 = titanic_df['Pclass'] == 1
cond3 = titanic_df['Sex'] == 'female'
titanic_df[ cond1 & cond2 & cond3 ]

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
275          276         1       1  Andrews, Mi...  female  63.0      1      0   13502  77.9583    D7        S
829          830         1       1  Stone, Mrs....  female  62.0      0      0  113572  80.0000   B28      NaN

### **8) 정렬, Aggregation 함수, GroupBy 적용**

#### **8-1) DataFrame, Series의 정렬 - sort_values() 메서드** 
주요 입력 파라미터는 by, ascending, inplace 
- by: 특정 column을 입력하면 **해당 column으로 정렬** 수행
- ascending: 정렬 방향 (Default: True)
  - True는 **오름차순**으로 정렬
  - False는 내림차순으로 정렬 
- inplace: 정렬 결과 적용 여부 (Default: False)
  - False는 **기존 DataFrame은 그대로 유지**하며 정렬된 DataFrame 결과로 반환
  - True는 호출한 DataFrame의 정렬 결과를 그대로 적용

**8-1-1) 오름차순 정렬**
- titanic_df를 Name column으로 오름차순 정렬

In [70]:
titanic_sorted = titanic_df.sort_values(by = ['Name'])
titanic_sorted.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch     Ticket   Fare Cabin Embarked
845          846         0       3  Abbing, Mr....    male  42.0      0      0  C.A. 5547   7.55   NaN        S
746          747         0       3  Abbott, Mr....    male  16.0      1      1  C.A. 2673  20.25   NaN        S
279          280         1       3  Abbott, Mrs...  female  35.0      1      1  C.A. 2673  20.25   NaN        S

**8-1-2) 여러 개의 column으로 정렬** <br/>
**by에 리스트 형식**으로 정렬하려는 column을 입력
- Pclass와 Name을 내림차순으로 정렬

In [72]:
titanic_sorted = titanic_df.sort_values(by = ['Pclass', 'Name'], ascending = False)
titanic_sorted.head(3)

PassengerId  Survived  Pclass            Name   Sex   Age  SibSp  Parch    Ticket  Fare Cabin Embarked
868          869         0       3  van Melkebe...  male   NaN      0      0    345777   9.5   NaN        S
153          154         0       3  van Billiar...  male  40.5      0      2  A/5. 851  14.5   NaN        S
282          283         0       3  de Pelsmaek...  male  16.0      0      0    345778   9.5   NaN        S

#### **8-2) Aggregation 함수 적용**
max(), min(), sum(), count()와 같은 aggregation 함수

**8-2-1) DataFrame에 aggregation 호출** <br/>
DataFrame에서 바로 aggregation을 호출할 경우 **모든 column에 해당 aggregation을 적용**
- 모든 column에 count() 결과 반환

In [73]:
titanic_df.count() 

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64

**8-2-2) 특정 column에 aggregation 함수 적용** <br/>
DataFrame에 **대상 column들만 추출**해 aggregation을 적용
- Age와 Fare에 mean() 결과 반환

In [74]:
titanic_df[['Age', 'Fare']].mean()

Age     29.699118
Fare    32.204208
dtype: float64

#### **8-3) groupby() 적용**
SQL과 판다스 모두 groupby를 분석 작업에 매우 많이 활용하는데, 서로 유사하면서도 다른 면이 존재
- DataFrame의 groupby() 사용 시 입력 파라미터 by에 column을 입력하면 **대상 column으로 groupby** 됨

**8-3-1) DataFrame GroupBy 객체** <br/>
DataFrame에 groupby() **호출 시** **DataFrameGroupBy라는 또 다른 형태의 DataFrame** 반환
- Pclass column 기준으로 GroupBy된 DataFrame Groupby 객체 반환

In [84]:
titanic_groupby = titanic_df.groupby(by = 'Pclass')
print(type(titanic_groupby))

<class 'pandas.core.groupby.generic.DataFrameGroupBy'>


**8-3-2) groupby() 후 aggregation 함수 적용** <br/>
DataFrame에 groupby()를 호출해 반환된 결과에 aggregation 함수 호출 시 <br/> **groupby() 대상 column을 제외한 모든 column에** 해당 aggregation 함수를 적용 
- groupby() 대상 column인 Pclass를 제외한 모든 column에 count 적용

In [85]:
titanic_groupby = titanic_df.groupby('Pclass').count()
titanic_groupby

PassengerId  Survived  Name  Sex  Age  SibSp  Parch  Ticket  Fare  Cabin  Embarked
Pclass                                                                                    
1               216       216   216  216  186    216    216     216   216    176       214
2               184       184   184  184  173    184    184     184   184     16       184
3               491       491   491  491  355    491    491     491   491     12       491

> **SQL에서 여러 개의 칼럼에 aggregation 함수 적용** <br/> 
> 여러 개의 칼럼에 aggregation 함수를 호출하려면 대상 column을 모두 select 절에 나열해야 함 <br/> 
>```
> # SQL
> Select count(PassengerID), count(Survived), ... from titanic_table group by Pclass
>```

**8-3-3) groupby()에 특정 칼럼만 aggregation 함수 적용** <br/>
groupby()로 반환된 DataFrameGroupBy 객체에 **해당 column을 필터링한 뒤 aggregation** 함수 적용

In [86]:
titanic_groupby = titanic_df.groupby('Pclass')[['PassengerId', 'Survived']].count()
titanic_groupby

PassengerId  Survived
Pclass                       
1               216       216
2               184       184
3               491       491

- titanic_df.groupby('Pclass')로 반환된 DataFrameGroupBy 객체에 **[[ 'PassengerID', 'Survived' ]]로 필터링**
- PassengerID와 Survived column에만 count() 수행

**8-3-4) groupby()에 서로 다른 aggregation 함수 적용** <br/>
적용하려는 **여러 개의 aggregation 함수명**을 DataFrameGroupBy 객체의 **agg() 내에 인자로 입력**해서 사용

In [87]:
titanic_df.groupby('Pclass')['Age'].agg([max, min])

max   min
Pclass            
1       80.0  0.92
2       70.0  0.67
3       74.0  0.42

> **SQL에서 서로 다른 aggregation 함수 적용** <br/> 
> 서로 다른 aggregation 함수를 적용할 경우에 Select 절에 나열 <br/> 
>```
> # SQL
> Select max(Age), min(Age), ... from titanic_table group by Pclass
>```

**8-3-5) groupby()에서 여러 개의 column에 서로 다른 aggregation 함수 적용** <br/>
agg() 인자 내에 입력 값으로 딕셔너리 형태로 aggregation이 적용될 column들과 aggregation 함수 입력
- DataFrame의 groupby()는 API기반으로 처리하므로 SQL보다 유연성이 떨어짐

In [88]:
agg_format = {'Age':'max', 'SibSp':'sum', 'Fare':'mean'}
titanic_df.groupby('Pclass').agg(agg_format)

Age  SibSp       Fare
Pclass                        
1       80.0     90  84.154687
2       70.0     74  20.662183
3       74.0    302  13.675550

### **9) 결손 데이터 처리하기**
결손 데이터(Missing Data)는 column에 값이 없는, NULL의 경우를 의미하며 넘파이의 NaN으로 표시
- 기본적으로 머신러닝 알고리즘은 이 NaN 값을 처리하지 않으므로 **다른 값으로 대체**
- NaN 값은 평균, 총합 등 함수 연산 시 제외 됨
- **NaN 여부를 확인하는 API는 isna(), NaN을 다른 값으로 대체하는 API는 fillna()**

### **9-1) isna()로 결손 데이터 여부 확인**
isna()를 이용 시 데이터가 NaN인지 아닌지 확인 가능

**9-1-1) DataFrame에 isna() 수행** <br/>
DataFrame에 isna() 수행 시 **모든 칼럼의 값이 NaN인지 아닌지를 True나 False로** 알려줌

In [89]:
titanic_df.isna().head(3)

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket   Fare  Cabin  Embarked
0        False     False   False  False  False  False  False  False   False  False   True     False
1        False     False   False  False  False  False  False  False   False  False  False     False
2        False     False   False  False  False  False  False  False   False  False   True     False

**9-1-2) 결손 데이터의 개수 구하기** <br/>
isna() 결과에 **sum() 함수를 추가**해 확인
- sum() 호출 시 True는 내부적으로 숫자 1, False는 숫자 0으로 변환되므로 결손 데이터의 개수 확인 가능

In [90]:
titanic_df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

#### **9-2) fillna()로 결손 데이터 대체**
fillna()를 이용 시 결손 데이터를 편리하게 다른 값으로 대체 가능
- 타이타닉 데이터 세트의 'Cabin' 칼럼의 NaN 값을 'C000'으로 대체

In [91]:
titanic_df['Cabin'] = titanic_df['Cabin'].fillna('C000')
titanic_df.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500  C000        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250  C000        S

**9-2-1) 실제 데이터 세트 값 변경** <br/>
fillna()를 통해 **반환 값을 다시 받거나 inplace = True 파라미터를 추가**해야 실제 데이터 세트 값이 변경 됨

In [92]:
# fillna() 반환 값을 다시 titanic_df['Cabin']으로 지정
titanic_df['Cabin'] = titanic_df['Cabin'].fillna('C000')

In [93]:
# inplace = True 사용
titanic_df['Cabin'].fillna('C000', inplace = True)

**9-2-2) 모든 결손 데이터 처리** <br/>
- 'Age' 칼럼의 NaN 값을 평균 나이로, 'Embarked' 칼럼의 NaN 값을 'S'로 대체

In [95]:
titanic_df['Age'] = titanic_df['Age'].fillna(titanic_df['Age'].mean())
titanic_df['Embarked'] = titanic_df['Embarked'].fillna('S')
titanic_df.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

### **10) apply lambda 식으로 데이터 가공**
apply 함수에 lamda 식을 결합해 DataFrame이나 Series의 레코드별로 데이터를 가공하는 기능 제공
- column에 **일괄적으로 데이터 가공을 하는 것이 속도면에서 더 빠르거나 복잡한 데이터 가공이 필요한 경우** 이용

#### **10-1) lambda 식**
파이썬에서 함수형 프로그래밍(functional programming)을 지원하기 위한 식

**10-1-1) lambda 식 예시**
- 입력값의 제곱 값을 구해서 반환하는 get_square(a) 함수
  - def get_square(a):와 같이 함수명과 입력 인자를 먼저 선언
  - 이후 함수 내에서 입력 인자를 가공한 뒤 결괏값을 return과 같은 문법으로 반환
- lambda 식을 통해 **함수의 선언과 함수 내의 처리를 한 줄의 식으로 쉽게 변환** 가능

In [96]:
def get_square(a):
  return a**2

print('3의 제곱은:', get_square(3))

3의 제곱은: 9


In [97]:
lambda_square = lambda x: x ** 2
print('3의 제곱은:', lambda_square(3))

3의 제곱은: 9


**10-1-2) lambda 식 기본 형태** <br/>
- ':'로 입력 인자와 반환될 입력 인자의 계산식 분리
  - ':' **왼쪽에 있는 x는 입력 인자**를 가리키며, **오른쪽은 입력 인자의 계산식**
  - 오른쪽의 계산식은 결국 **반환 값**을 의미
- 여러 개의 값을 입력 인자로 사용해야할 경우에는 **map() 함수**를 결합하여 사용
  - map(적용시킬 함수, 적용할 값들)
  - 두번째 인자로는 리스트, 튜플 등의 반복 가능한 자료형이 들어와야 함
  - map 함수의 반환 값은 map객체 이기 때문에 **해당 자료형을 list 혹은 tuple로 형 변환**

In [99]:
a = [1, 2, 3]
squares = map(lambda x : x**2, a)
list(squares)

[1, 4, 9]

#### **10-2) DataFrame의 apply에 lambda 식을 적용해 데이터 가공**

**10-2-1) 데이터 가공 예시**
- 'Name' column의 문자열 개수를 별도의 column인 'Name_len'에 생성

In [100]:
titanic_df['Name_len'] = titanic_df['Name'].apply(lambda x : len(x))
titanic_df[['Name', 'Name_len']].head(3)

Name  Name_len
0  Braund, Mr....        23
1  Cumings, Mr...        51
2  Heikkinen, ...        22

**10-2-2) if else를 이용한 데이터 가공 예시** <br/>
- if 절의 경우 **if 식보다 반환 값을 먼저 기술**해야 함
  - lamda 식 ':' 기호의 오른편에 반환 값이 있어야 하기 때문
  - ex. lambda x : 'Child' if x <= 15 else 'Adult'
- else의 경우 **else 식이 먼저 나오고 반환 값이 나중에** 오면 됨

In [101]:
titanic_df['Child_Adult'] = titanic_df['Age'].apply(lambda x : 'Child' if x <= 15 else 'Adult')
titanic_df[['Age', 'Child_Adult']].head(8)

Age Child_Adult
0  22.000000       Adult
1  38.000000       Adult
2  26.000000       Adult
3  35.000000       Adult
4  35.000000       Adult
5  29.699118       Adult
6  54.000000       Adult
7   2.000000       Child

- 나이가 15세 미만이면 'Child', 그렇지 않으면 'Adult'로 구분하는 새로운 column 'Child_Adult' 생성

**10-2-3) else if을 이용한 데이터 가공 예시** <br/>
if, else만 지원하고 if, else if, else와 같이 else if는 지원하지 않음
- else if를 이용하기 위해서는 **else 절을 ( )로 내포해 ( ) 내에서 다시 if else 적용**
- () 안에서도 if 절의 경우 반환 값이 식보다 먼저 나왔음에 유의

In [104]:
titanic_df['Age_cat'] = titanic_df['Age'].apply(lambda x : 'Child' if x <= 15 else ('Adult' if x <= 60 else 'Elderly'))

titanic_df['Age_cat'].value_counts()

Adult      786
Child       83
Elderly     22
Name: Age_cat, dtype: int64

- 나이가 15세 이하면 Child, 15~60세 사이는 Adult, 61세 이상은 Elderly로 분류하는 'Age_Cat' Column 생성

**10-2-4) 세분화 된 분류 예시** <br/>
else if가 많이 나와야 하는 경우나 switch case 문의 경우 else를 계속 내포해서 사용하기 부담스러움
- 세분화 된 분류의 경우 별도의 함수를 사용하는 것이 바람직

In [105]:
# 나이에 따라 세분화된 분류를 수행하는 함수 생성.
def get_category(age):
  cat = ''
  if age <= 5: cat = 'Baby'
  elif age <= 12: cat = 'Child'
  elif age <= 25: cat = 'Teenager'
  elif age <= 35: cat = 'Young Adult'
  elif age <= 60: cat = 'Adult'
  else: cat = 'Elderly'
  
  return cat

# lambda 식에 위에서 생성한 get_category() 함수를 반환값으로 지정.
# get_category(X)는 입력값으로 'Age' 칼럼 값을 받아서 해당하는 cat 반환
titanic_df['Age_cat'] = titanic_df['Age'].apply(lambda x : get_category(x))
titanic_df[['Age', 'Age_cat']].head()

Age      Age_cat
0  22.0     Teenager
1  38.0        Adult
2  26.0  Young Adult
3  35.0  Young Adult
4  35.0  Young Adult

- 5세 이하는 Baby, 12살 이하는 Child, 18살 이하는 Teenage, 25살 이하는 Student, <br/> 35살 이하는 Young Adult, 60세 이하는 Adult, 그 이상은 Elderly로 분류하는 함수 생성